src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+
+  const circpad_time_t histogram_range_usec =
+                       state->range_sec*TOR_USEC_PER_SEC;
+  bin = state->histogram_len -
+    tor_log2(histogram_range_usec/(us-start_usec+1));
+
+  /* Clamp the return value to account for timevals before the start
+   * of bin 0, or after the last bin. Don't return the infinity bin
+   * index. */
+  bin = MIN(MAX(bin, 1), state->histogram_len-2);
+  return bin;
+}
+
+/**
+ * This function frees any token bins allocated from a previous state
+ *
+ * Called after a state transition, or if the bins are empty.
+ */
+STATIC void
+circpad_machine_setup_tokens(circpad_machineinfo_t *mi)
+{
+  const circpad_state_t *state = circpad_machine_current_state(mi);
+
+  /* If this state doesn't exist, or doesn't have token removal,
+   * free any previous state's histogram, and bail */
+  if (!state || state->token_removal == CIRCPAD_TOKEN_REMOVAL_NONE) {
+    if (mi->histogram) {
+      tor_free(mi->histogram);
+      mi->histogram = NULL;
+      mi->histogram_len = 0;
+    }
+    return;
+  }
+
+  /* Try to avoid re-mallocing if we don't really need to */
+  if (!mi->histogram || (mi->histogram
+          && mi->histogram_len != state->histogram_len)) {
+    tor_free(mi->histogram); // null ok
+    mi->histogram = tor_malloc_zero(sizeof(circpad_hist_bin_t)
+                                    *state->histogram_len);
+  }
+  mi->histogram_len = state->histogram_len;
+
+  memcpy(mi->histogram, state->histogram,
+         sizeof(circpad_hist_bin_t)*state->histogram_len);
+}
+
+/**
+ * Choose a length for this state (in cells), if specified.
+ */
+static void
+circpad_choose_state_length(circpad_machineinfo_t *mi)
+{
+  const circpad_state_t *state = circpad_machine_current_state(mi);
+  double length;
+
+  if (!state || state->length_dist.type == CIRCPAD_DIST_NONE) {
+    mi->state_length = CIRCPAD_STATE_LENGTH_INFINITE;
+    return;
+  }
+
+  length = circpad_distribution_sample(state->length_dist);
+  length = MAX(0, length);
+  length += state->start_length;
+  length = MIN(length, state->max_length);
+
+  mi->state_length = clamp_double_to_int64(length);
+}
+
+/**
+ * Sample an expected time-until-next-packet delay from the histogram.
+ *
+ * The bin is chosen with probability proportional to the number
+ * of tokens in each bin, and then a time value is chosen uniformly from
+ * that bin's [start,end) time range.
+ */
+static circpad_delay_t
+circpad_machine_sample_delay(circpad_machineinfo_t *mi)
+{
+  const circpad_state_t *state = circpad_machine_current_state(mi);
+  const circpad_hist_bin_t *histogram = NULL;
+  int i = 0;
+  uint32_t curr_weight = 0;
+  uint32_t histogram_total = 0;
+  uint32_t bin_choice;
+  circpad_delay_t bin_start, bin_end;
+  circpad_delay_t start_usec;
+
+  tor_assert(state);
+
+  if (state->use_rtt_estimate)
+    start_usec = mi->rtt_estimate_us+state->start_usec;
+  else
+    start_usec = state->start_usec;
+
+  if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE) {
```

Comment:
This is a weird if block because its branches are disconnected.
IIUC, it checks whether the histogram is immutable or mutable, and it also checks if we are using a histogram or a distribution in the same if block, and if it's a distribution it samples it on the spot.

Might be nicer to do this cleanly and first check if we are using a distribution, and handle that (ideally in its own function). And then if we are using a histogram, check if it's mutable or immutable. instead of doing everything at the same time.

Reply:
22fbd5fa796c483bda6e10b0015c9feedf288300